src/hostapi/oboe/pa_oboe.cpp

```diff
@@ -0,0 +1,1900 @@
+            mediator->setOutputCallback();
+        }
+
+        result = mediator->m_outputBuilder.openStream(mediator->m_outputStream);
+        if (result != Result::OK) {
+            LOGE("[OboeEngine::OpenStream]\t Oboe couldn't open the output stream: %s",
+                 convertToText(result));
+            return paUnanticipatedHostError;
+        }
+
+        mediator->m_outputStream->setBufferSizeInFrames(mediator->m_outputStream->getFramesPerBurst() *
+                                                          g_numberOfBuffers);
+        i_oboeStream->outputBuffers =
+                (void **) PaUtil_AllocateZeroInitializedMemory(g_numberOfBuffers * sizeof(int32_t * ));
+
+        for (int i = 0; i < g_numberOfBuffers; ++i) {
+            i_oboeStream->outputBuffers[i] = (void *) PaUtil_AllocateZeroInitializedMemory(
+                    i_oboeStream->framesPerHostCallback *
+                    i_oboeStream->bytesPerFrame *
+                    i_oboeStream->bufferProcessor.outputChannelCount);
+
+            if (!i_oboeStream->outputBuffers[i]) {
+                for (int j = 0; j < i; ++j)
+                    PaUtil_FreeMemory(i_oboeStream->outputBuffers[j]);
+                PaUtil_FreeMemory(i_oboeStream->outputBuffers);
+                mediator->m_outputStream->close();
+                error = paInsufficientMemory;
+                break;
+            }
+        }
+        i_oboeStream->currentOutputBuffer = 0;
+    }
+
+    return error;
+}
+
+
+/**
+ * \brief   Starts oboeStream - both input and output AudioStreams of the OboeStream are checked
+ *          and requested to be started.
+ * @param   oboeStream The stream we want to start.
+ * @return  true if the streams we wanted to start are started successfully, false otherwise.
+ */
+bool OboeEngine::startStream(OboeStream *i_oboeStream) {
+    Result outputResult = Result::OK, inputResult = Result::OK;
+    OboeMediator* mediator = i_oboeStream->oboeMediator;
+
+    if (i_oboeStream->hasInput) {
+        inputResult = mediator->m_inputStream->requestStart();
+        if (inputResult != Result::OK)
+            LOGE("[OboeEngine::startStream]\t Oboe couldn't start the input stream: %s",
+                 convertToText(inputResult));
+    }
+    if (i_oboeStream->hasOutput) {
+        outputResult = mediator->m_outputStream->requestStart();
+        if (outputResult != Result::OK)
+            LOGE("[OboeEngine::startStream]\t Oboe couldn't start the output stream: %s",
+                 convertToText(outputResult));
+    }
+
+    return (outputResult == Result::OK && inputResult == Result::OK);
+}
+
+
+/**
+ * \brief   Stops oboeStream - both input and output AudioStreams of the OboeStream are checked
+ *          and requested to be stopped.
+ * @param   oboeStream The stream we want to stop.
+ * @return  true if the streams we wanted to stop are stopped successfully, false otherwise.
+ */
+bool OboeEngine::stopStream(OboeStream *i_oboeStream) {
```

Comment:
We have been moving to passing shared_ptr for streams and callbacks as much as possible.
It avoids many crashes. Would that be easy to do for some of these methods?

Reply:
Do you mean changing each pointer I use in the OboeEngine and OboeMediator classes? That is possible, however I think I cannot make any OboeStream pointer a shared one, because OboeStream is the extention of PaStream, a C struct, and, in my limited experience, I noticed that having shared pointers in that struct will cause a segmentation fault - that's the reason why I made the OboeMediator class in the first place.

Reply:
@philburk Good morning, any chance I can get any clarification on this matter?